File: ExpressControls/MenuButton.vb
' Follow steps 1a or 1b and then 2 to use this custom control in a XAML file.
'
' Step 1a) Using this custom control in a XAML file that exists in the current project.
' Add this XmlNamespace attribute to the root element of the markup file where it is 
' to be used:
'
'     xmlns:MyNamespace="clr-namespace:ExpressControls"
'
'
' Step 1b) Using this custom control in a XAML file that exists in a different project.
' Add this XmlNamespace attribute to the root element of the markup file where it is 
' to be used:
'
'     xmlns:MyNamespace="clr-namespace:ExpressControls;assembly=ExpressControls"
'
' You will also need to add a project reference from the project where the XAML file lives
' to this project and Rebuild to avoid compilation errors:
'
'     Right click on the target project in the Solution Explorer and
'     "Add Reference"->"Projects"->[Browse to and select this project]
'
'
' Step 2)
' Go ahead and use your control in the XAML file. Note that Intellisense in the
' XML editor does not currently work on custom controls and its child elements.
'
'     <MyNamespace:MenuButton/>
'

Public Class MenuButton
    Inherits Button

    Shared Sub New()
        'This OverrideMetadata call tells the system that this element wants to provide a style that is different than its base class.
        'This style is defined in themes\generic.xaml
        DefaultStyleKeyProperty.OverrideMetadata(GetType(MenuButton), New FrameworkPropertyMetadata(GetType(MenuButton)))
    End Sub

    Public Shared ReadOnly IconProperty As DependencyProperty = DependencyProperty.Register(NameOf(Icon), GetType(Viewbox), GetType(MenuButton), New PropertyMetadata(Nothing))

    Public Property Icon As Viewbox
        Get
            Return CType(GetValue(IconProperty), Viewbox)
        End Get
        Set(ByVal value As Viewbox)
            SetValue(IconProperty, value)
        End Set
    End Property

    Public Shared ReadOnly TextProperty As DependencyProperty = DependencyProperty.Register(NameOf(Text), GetType(String), GetType(MenuButton), New PropertyMetadata(Nothing))

    Public Property Text As String
        Get
            Return CStr(GetValue(TextProperty))
        End Get
        Set(ByVal value As String)
            SetValue(TextProperty, value)
        End Set
    End Property

    Public Shared ReadOnly TextVisibilityProperty As DependencyProperty = DependencyProperty.Register(NameOf(TextVisibility), GetType(Visibility), GetType(MenuButton), New PropertyMetadata(Nothing))

    Public Property TextVisibility As Visibility
        Get
            Return CType(GetValue(TextVisibilityProperty), Visibility)
        End Get
        Set(ByVal value As Visibility)
            SetValue(TextVisibilityProperty, value)
        End Set
    End Property

    Public Shared ReadOnly MoreVisibilityProperty As DependencyProperty = DependencyProperty.Register(NameOf(MoreVisibility), GetType(Visibility), GetType(MenuButton), New PropertyMetadata(Nothing))

    Public Property MoreVisibility As Visibility
        Get
            Return CType(GetValue(MoreVisibilityProperty), Visibility)
        End Get
        Set(ByVal value As Visibility)
            SetValue(MoreVisibilityProperty, value)
        End Set
    End Property

    Public Shared ReadOnly IsMiniProperty As DependencyProperty = DependencyProperty.Register(NameOf(IsMini), GetType(Boolean), GetType(MenuButton), New FrameworkPropertyMetadata(AddressOf IsMiniPropertyChanged))

    Public Property IsMini As Boolean
        Get
            Return CBool(GetValue(IsMiniProperty))
        End Get
        Set(ByVal value As Boolean)
            SetValue(IsMiniProperty, value)
        End Set
    End Property

    Public Shared Sub IsMiniPropertyChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
        Dim b As Boolean = CBool(e.NewValue)
    End Sub

    Public Shared ReadOnly IsLSquaredProperty As DependencyProperty = DependencyProperty.Register(NameOf(IsLSquared), GetType(Boolean), GetType(MenuButton), New FrameworkPropertyMetadata(AddressOf IsLSquaredPropertyChanged))

    Public Property IsLSquared As Boolean
        Get
            Return CBool(GetValue(IsLSquaredProperty))
        End Get
        Set(ByVal value As Boolean)
            SetValue(IsLSquaredProperty, value)
        End Set
    End Property

    Public Shared Sub IsLSquaredPropertyChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
        Dim b As Boolean = CBool(e.NewValue)
    End Sub

    Public Shared ReadOnly IsRSquaredProperty As DependencyProperty = DependencyProperty.Register(NameOf(IsRSquared), GetType(Boolean), GetType(MenuButton), New FrameworkPropertyMetadata(AddressOf IsRSquaredPropertyChanged))

    Public Property IsRSquared As Boolean
        Get
            Return CBool(GetValue(IsRSquaredProperty))
        End Get
        Set(ByVal value As Boolean)
            SetValue(IsRSquaredProperty, value)
        End Set
    End Property

    Public Shared Sub IsRSquaredPropertyChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
        Dim b As Boolean = CBool(e.NewValue)
    End Sub

    Public Shared ReadOnly IsUpFacingMenuProperty As DependencyProperty = DependencyProperty.Register(NameOf(IsUpFacingMenu), GetType(Boolean), GetType(MenuButton), New FrameworkPropertyMetadata(AddressOf IsUpFacingMenuPropertyChanged))

    Public Property IsUpFacingMenu As Boolean
        Get
            Return CBool(GetValue(IsUpFacingMenuProperty))
        End Get
        Set(ByVal value As Boolean)
            SetValue(IsUpFacingMenuProperty, value)
        End Set
    End Property

    Public Shared Sub IsUpFacingMenuPropertyChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
        Dim b As Boolean = CBool(e.NewValue)
    End Sub

End Class

File: ExpressControls/MenuButton.cs
using System;
using System.Windows;
using System.Windows.Controls;

namespace ExpressControls
{
    /// <summary>
    /// Follow steps 1a or 1b and then 2 to use this custom control in a XAML file.
    ///
    /// Step 1a) Using this custom control in a XAML file that exists in the current project.
    /// Add this XmlNamespace attribute to the root element of the markup file where it is
    /// to be used:
    ///
    ///     xmlns:MyNamespace="clr-namespace:ExpressControls"
    ///
    ///
    /// Step 1b) Using this custom control in a XAML file that exists in a different project.
    /// Add this XmlNamespace attribute to the root element of the markup file where it is
    /// to be used:
    ///
    ///     xmlns:MyNamespace="clr-namespace:ExpressControls;assembly=ExpressControls"
    ///
    /// You will also need to add a project reference from the project where the XAML file lives
    /// to this project and Rebuild to avoid compilation errors:
    ///
    ///     Right click on the target project in the Solution Explorer and
    ///     "Add Reference"->"Projects"->[Select this project]
    ///
    ///
    /// Step 2)
    /// Go ahead and use your control in the XAML file.
    ///
    ///     <MyNamespace:CustomControl1/>
    ///
    /// </summary>
    public class MenuButton : Button
    {
        static MenuButton()
        {
            DefaultStyleKeyProperty.OverrideMetadata(
                typeof(MenuButton),
                new FrameworkPropertyMetadata(typeof(MenuButton))
            );
        }

        public static readonly DependencyProperty IconProperty = DependencyProperty.Register(
            nameof(Icon),
            typeof(Viewbox),
            typeof(MenuButton),
            new PropertyMetadata(null)
        );

        public Viewbox Icon
        {
            get { return (Viewbox)GetValue(IconProperty); }
            set { SetValue(IconProperty, value); }
        }

        public static readonly DependencyProperty TextProperty = DependencyProperty.Register(
            nameof(Text),
            typeof(string),
            typeof(MenuButton),
            new PropertyMetadata(null)
        );

        public string Text
        {
            get { return Convert.ToString(GetValue(TextProperty)) ?? ""; }
            set { SetValue(TextProperty, value); }
        }

        public static readonly DependencyProperty TextVisibilityProperty =
            DependencyProperty.Register(
                nameof(TextVisibility),
                typeof(Visibility),
                typeof(MenuButton),
                new PropertyMetadata(null)
            );

        public Visibility TextVisibility
        {
            get { return (Visibility)GetValue(TextVisibilityProperty); }
            set { SetValue(TextVisibilityProperty, value); }
        }

        public static readonly DependencyProperty MoreVisibilityProperty =
            DependencyProperty.Register(
                nameof(MoreVisibility),
                typeof(Visibility),
                typeof(MenuButton),
                new PropertyMetadata(null)
            );

        public Visibility MoreVisibility
        {
            get { return (Visibility)GetValue(MoreVisibilityProperty); }
            set { SetValue(MoreVisibilityProperty, value); }
        }

        public static readonly DependencyProperty IsMiniProperty = DependencyProperty.Register(
            nameof(IsMini),
            typeof(bool),
            typeof(MenuButton),
            new FrameworkPropertyMetadata(IsMiniPropertyChanged)
        );

        public bool IsMini
        {
            get { return Convert.ToBoolean(GetValue(IsMiniProperty)); }
            set { SetValue(IsMiniProperty, value); }
        }

        public static void IsMiniPropertyChanged(
            DependencyObject d,
            DependencyPropertyChangedEventArgs e
        ) { }

        public static readonly DependencyProperty IsLSquaredProperty = DependencyProperty.Register(
            nameof(IsLSquared),
            typeof(bool),
            typeof(MenuButton),
            new FrameworkPropertyMetadata(IsLSquaredPropertyChanged)
        );

        public bool IsLSquared
        {
            get { return Convert.ToBoolean(GetValue(IsLSquaredProperty)); }
            set { SetValue(IsLSquaredProperty, value); }
        }

        public static void IsLSquaredPropertyChanged(
            DependencyObject d,
            DependencyPropertyChangedEventArgs e
        ) { }

        public static readonly DependencyProperty IsRSquaredProperty = DependencyProperty.Register(
            nameof(IsRSquared),
            typeof(bool),
            typeof(MenuButton),
            new FrameworkPropertyMetadata(IsRSquaredPropertyChanged)
        );

        public bool IsRSquared
        {
            get { return Convert.ToBoolean(GetValue(IsRSquaredProperty)); }
            set { SetValue(IsRSquaredProperty, value); }
        }

        public static void IsRSquaredPropertyChanged(
            DependencyObject d,
            DependencyPropertyChangedEventArgs e
        ) { }

        public static readonly DependencyProperty IsUpFacingMenuProperty =
            DependencyProperty.Register(
                nameof(IsUpFacingMenu),
                typeof(bool),
                typeof(MenuButton),
                new FrameworkPropertyMetadata(IsUpFacingMenuPropertyChanged)
            );

        public bool IsUpFacingMenu
        {
            get { return Convert.ToBoolean(GetValue(IsUpFacingMenuProperty)); }
            set { SetValue(IsUpFacingMenuProperty, value); }
        }

        public static void IsUpFacingMenuPropertyChanged(
            DependencyObject d,
            DependencyPropertyChangedEventArgs e
        ) { }
    }
}
